store/sqlstore/channel_store.go

```diff
@@ -416,6 +421,260 @@ func (s SqlChannelStore) createIndexesIfNotExists() {
+	sb := s.
+		getQueryBuilder().
+		Select("Preferences.UserId", "Preferences.Name AS ChannelId", "SidebarCategories.Id AS CategoryId").
+		From("Preferences").
+		Where(sq.And{
+			sq.Eq{"Preferences.Category": model.PREFERENCE_CATEGORY_FAVORITE_CHANNEL},
+			sq.NotEq{"SidebarCategories.Id": nil},
+			sq.Gt{"Preferences.UserId": lastUserId},
+		}).
+		LeftJoin("Channels ON (Channels.Id=Preferences.Name)").
+		LeftJoin("SidebarCategories ON (SidebarCategories.UserId=Preferences.UserId AND SidebarCategories.TeamId=Channels.TeamId AND SidebarCategories.Type='F')").
+		OrderBy("Preferences.UserId")
+
+	transaction, err := s.GetMaster().Begin()
+	if err != nil {
+		return nil, model.NewAppError("SqlChannelStore.MigrateFavoritesToSidebarChannels", "store.sql_channel.migrate_favorites.open_transaction.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	defer finalizeTransaction(transaction)
+
+	sql, args, err := sb.ToSql()
+	if err != nil {
+		return nil, model.NewAppError("SqlChannelStore.MigrateFavoritesToSidebarChannels", "store.sql_channel.migrate_favorites.select.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	var userFavorites []userMembership
+	if _, err := transaction.Select(&userFavorites, sql, args...); err != nil {
+		return nil, model.NewAppError("SqlTeamStore.MigrateFavoritesToSidebarChannels", "store.sql_channel.migrate_favorites.select.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	for _, favorite := range userFavorites {
+		sql, args, _ := s.getQueryBuilder().
+			Insert("SidebarChannels").
+			Columns("ChannelId", "UserId", "CategoryId", "SortOrder").
+			Values(favorite.ChannelId, favorite.UserId, favorite.CategoryId, 0).ToSql()
+
+		if _, err := transaction.Exec(sql, args...); err != nil && !IsUniqueConstraintError(err, []string{"UserId"}) {
+			return nil, model.NewAppError("SqlChannelStore.MigrateFavoritesToSidebarChannels", "store.sql_channel.migrate_favorites.open_transaction.app_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+	}
+
+	if err := transaction.Commit(); err != nil {
+		return nil, model.NewAppError("SqlChannelStore.MigrateFavoritesToSidebarChannels", "store.sql_channel.migrate_favorites.commit_transaction.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	data := make(map[string]string)
+	data["UserId"] = userFavorites[len(userFavorites)-1].UserId
+
+	return data, nil
+
+}
+
+func (s SqlChannelStore) MigrateChannelsToSidebarChannels(lastChannelId, lastUserId string) (map[string]string, *model.AppError) {
+	sb := s.
+		getQueryBuilder().
+		Select("Channels.Id AS ChannelId", "ChannelMembers.UserId", "SidebarCategories.Id AS CategoryId").
+		From("Channels").
+		Where(sq.And{
+			sq.Or{
+				sq.Eq{"Channels.Type": model.CHANNEL_OPEN},
+				sq.Eq{"Channels.Type": model.CHANNEL_PRIVATE},
+			},
+			sq.NotEq{"SidebarCategories.Id": nil},
+			sq.Gt{"Channels.Id": lastChannelId, "ChannelMembers.UserId": lastUserId},
```

Comment:
One invocation of this function migrates one DM channel? Seems like this might take a while? Again, maybe the batch size should be a bit larger?